### ISSUE TITLE
rendering infrastructure overhaul

### ISSUE THREAD

**Issue by Aeva** [Aeva/m.grl#232]
M.GRL performs very poorly when a modest number of dynamic objects are in play at once.  This is however a fixable problem, and I have a much better idea as to what kind of rendering architecture is needed to support this now than I did when I started down the rabbit hole of "adding WebGL" a few years ago.

The following tickets contain relevant exploratory research on how to resolve the shortcomings of the engine: #189 (consolidated VBOs), #193 (related stress test), #195 (using instancing extension instead), #197 (glsl compiler support for instancing abstraction).

The most striking part of pursuing the instancing extension was one of my "control group" tests: just flattening out the rendering graph to draw as many of one object as possible.  No extensions, no VBO foolery, just no branching what-so-ever.  The results were good enough (though it seems I didn't feel the need to document them anywhere obvious) that it became evident that I needed to employ code generation for any execution pathway that is time sensitive.  The rendering bottleneck is entirely CPU bound, so that should be made to run as fast as possible, and that means removing unnecessary work at draw time.

**Comment by Aeva** [comment]
This'll probably keep me busy until September XD

**Comment by Aeva** [comment]
Good progress this weekend - the fast_graph branch now has JTA instances using code generation for the draw paths.  Without applying things like state sorting or having the entire draw path as a generated function, performance seems to be about the same.  If anything, it feels a like it gets up to speed faster, which is encouraging, though I don't have a good way to measure this.  Average frame rate is about the same, which is good given that this is the unoptimized form.

This path way is a lot stricter about binding contexts - if a uniform isn't marked as such, then it doesn't get an IR in the codegen path.  This breaks most of the demos.  I think once I have a regen mechanism in place, that same path could be used to add more uniform bindings as needed.

This is what I think needs to be done next:

 - [ ] an ir_dirty event on graph nodes - this should be triggered whenever a '.shader' driver property is set
 - [ ] use the above to allow IR to change from static to dynamic
 - [ ] use the above to trigger the __static_draw path to be replaced
 - [ ] use the above to trigger more IR objects to be added to the __static_draw path

Once this is done, the demos seem to no longer be broken due to overly strict binding context behavior then I think the next step would be to start on creating a monolithic draw function.

**Comment by Aeva** [comment]
Ok, this seems to work really great now.  The location_picking demo now runs at 60fps without using any fancy psuedo-static modes.

 - [ ] use the dirty signals to trigger more IR objects to be added to the __static_draw path
 - [ ] remove __static_draw, replace with just a per-node src cache & generate a monolithic static function on the graph root

**Comment by Aeva** [comment]
- [ ] use the dirty signals to trigger more IR objects to be added to the __static_draw path (?)
 - [ ] fix regression where properties that should end up static aren't (see 'mode' in demo 6)
 - [ ] update unit tests for codegen

**Comment by Aeva** [comment]
Some thoughts for where to go next:


I think it might make sense to put
together some basic state sorting while keeping in mind that I should
put together the picking pass stuff next.

 1. the "ir list" that is generated in m.codegen.js (called by
    m.jta.js) and consumed by m.graph.js should be replaced with an
    object with some properties that we can use for sorting (vbo_id,
    ibo_id, current static sampler bindings, etc)

 2. the object should be able to compile to one of several "profiles".
    So far, this is probably just "generic" and "picking"

 3. throw together some rudimentrary state sorting.  I think sort for
    VBO, IBO, static samplers, and then static unis

 4. eliminate redundant static calls, ideally on the fly


It would be good if we could eliminate some redundancy by doing the
picking stuff in tandem with the generic draw pass.  I think this
would work ok, because the picking pass is just a uniform and the
binds, so the more constrained sorting could just be re-used.

**Comment by Aeva** [comment]
Ok, the graph draw function is now generated by the RenderNode (which has a 1:1 relationship with a shader program) rather than by the SceneGraph (who's objects can be associated to many RenderNodes and shader programs).

I added a lispy 'quote' macro, so the entire static draw function is just generated in one go and replaces the RenderNode's 'render' method.

As mentioned in commit a51d553, there are some places where I feel this is half-assed and could use revision:

 - [ ] Cannot use cpp macros inside the quote macro.  We might benefit from a different syntax that cpp doesn't trip over.  Ideally we could run cpp first.  Considering trying the QUOTE {...} syntax instead.

 - [ ] Intelligent whitespace would be nice in the generated quoted regions.

 - [ ] commented out some "exclude test" stuff, probably should just remove it entirely

 - [ ] dummied out SceneGraph's "__create_picking_node" method because it causes problems as is.  Could be fixed by running it out of phase via set timeout, so that it doesn't try to compile the graph node before the graph node finished initializing...?

 - [ ] a mechanism for swapping out the shader program on the RenderNode would be nice.  it should also trigger a recompile event so that the code is accurate.

**Comment by Aeva** [comment]
also note the following demos are still broken, but should be easier to resolve now:
 - [ ] site demo doesn't work (missing attribute 'normals') [1]
 - [ ] compositing demo doesn't work (missing attribute 'tcoords') [1]
 - [ ] particle system stops rendering after transition effect (not sure why, though there is a GL error)
 - [ ] glsl compiler demo doesn't work at all (needs to be able late derive uniform bindings maybe?)

**Comment by Aeva** [comment]
Currently the jta resources generate their IR once, at load time, with respect to the currently active shader, and never again.  The main problem this results in, is that the attributes and uniforms won't necessarily match what is needed to draw under the desired shader program.

The ideal case might instead be an IR object, which puts together the initial bindings, but is able to dynamically add new bindings as needed.  Then, when it is time to compile, a method would be called to return the specific IR list needed to draw the object in the current shader.

**Comment by Aeva** [comment]
Currently the location picking demo in master is faster (60fsp average on the x200) than the one in fast_graph (40fps).  However, the one in master has some hacks to make some drawables immutable, whereas the tiles in fast_graph are all dynamic (so they call and upload 4 driver functions each before drawing)

**Comment by Aeva** [comment]
I added a "freeze" method to GraphNode instances, which causes them to be drawn entirely as static objects.  The performance on location_picking shoots up to 60hz in testing with it, giving equivalent or better performance to for static drawing but with a much nicer API and memory usage footprint.

**Comment by Aeva** [comment]
Current regressions in demos:
 - [ ] light demo is a blank screen
 - [ ] compositing demo's pip is just a solid color
 - [ ] particle demo's particles don't render after loading screen transition effect ends
 - [ ] anything relying on picking is broken

**Comment by Aeva** [comment]
Basic outline for re-implementing object picking:

 1. PickingNode singleton
    
    - low resolution buffer
    - use the draw function generator from RenderNode
    - all nodes considered pickable, for now (except particles)
    - shader should limit what properties need to be set
    - coerce GarphNode object_ids to always be static
    - picking index should work a little differently: statics should
      be given incremented IDs upon the picking pass being compiled.
      Non statics should be enumerated in gl_tick as currently, but
      not starting from 0

 2. new picking pass shader (mrt, floating point textures  etc)

 3. connect to existing API

 4. remove dead code from old picking system

 5. test, debug, etc

 6. burn-in mechansim?

 7. "classic" variation that doesn't use mrt or floating point buffers

**Comment by Aeva** [comment]
Got sidetracked exploring building a simpler case study for codegen as an optimization for a blog post that didn't play out as expected.  Which is fine I guess, as it helped clarify how mgrl's NOGL renderer should work.

In the mean time, I've been puzzling over what would be the best to do with alpha blending, and noted some things in the ticket's description above.  Might continue with either that or picking next.  The two have some similarities though, so it might be good for me to sleep it over a little more and figure out what infrastructure can support both best.

**Comment by Aeva** [comment]
Moved reimplementing the picking system out into issue #237 to keep notes tidy.

**Comment by Aeva** [comment]
merged pick_nouveau into fast_graph

**Comment by Aeva** [comment]
Was feeling a desire to work on features instead of fixing things =) so, the last few commits implement an API for "stamping" assets into a SceneGraph instance without actually creating individual GraphNodes for them.  The new "static_drawing" demo demonstrates the API for this.  For drawing ~10200 objects, this isn't as fast as it could be running.  The problem is obvious when looking at the generated rendering code, where you get a few thousand lines of array declaration syntax... oops.

So, this could be improved by maybe instead, generating and storing a singular long float array at compile time, and referencing sections of that instead.

This could probably be accomplished by tracking the graph ID of the stamp, and storing the IR in an object instead of an array, allowing for the space needed to be allocated at compile time.

**Comment by Aeva** [comment]
Ok, new branch "heap_experiment" tracks trying to speed up the static drawing demo.  It doesn't work right now though, for reasons which are detailed in the last commit message.  Considering forking this off into its own issue, but gist is that the JS IR objects created for uniform uploads should be replaced with IR that actually just produces the actual GL call.  From what I've been able to figure out, that will cause the 'new Float32Array' stuff in the complied output to basically function as a type cast instead of a new object.  God damn it, javascript >_>

**Comment by Aeva** [comment]
Closed out issue #238, and merged the resulting branch back into fast_graph.

**Comment by Aeva** [comment]
Only thing that is overtly broken at this point is the lighting demo (#244)

**Comment by Aeva** [comment]
Did some profiling tonight, since I discovered google chrome's profiler doesn't have any problems with the generated functions, unlike firefox developer edition.  Right now, one of the most expensive things is the "__regen_glsl_bindings" stuff, which is something I want to get rid of anyway.

The actual generated rendering functions occupy a very percentage of the run time (about 5%)!!!  Most of the demos run at 60fps (with a slight slow down for mouse picking).

While this may very well be due to improvements to chrome, I am feeling somewhat emboldened that the last year of development was not a total waste afterall <3

**Comment by Aeva** [comment]
Merged the ```instancing``` branch into ```fast_graph``` <3
